src/sp/it/pl/gui/objects/contextmenu/ContextMenuGenerator.kt

```diff
@@ -0,0 +1,85 @@
+package sp.it.pl.gui.objects.contextmenu
+
+import javafx.event.ActionEvent
+import javafx.scene.Node
+import javafx.scene.control.ContextMenu
+import javafx.scene.control.Menu
+import javafx.scene.control.MenuItem
+import sp.it.pl.gui.pane.collectionUnwrap
+import sp.it.pl.gui.pane.collectionWrap
+import sp.it.pl.util.collections.map.ClassListMap
+import sp.it.pl.util.dev.fail
+import sp.it.pl.util.functional.asArray
+import sp.it.pl.util.functional.asIf
+import sp.it.pl.util.functional.getElementType
+import sp.it.pl.util.functional.net
+import sp.it.pl.util.functional.seqOf
+import sp.it.pl.util.graphics.menuItem
+import sp.it.pl.util.graphics.menuSeparator
+
+val contextMenuGenerator = ContextMenuGenerator()
+
+/** Context menu generator with per type generator registry */
+class ContextMenuGenerator {
+    private val mSingle = ClassListMap<(ContextMenu, Any?) -> Sequence<MenuItem>> { fail() }
+    private val mMany = ClassListMap<(ContextMenu, Any?) -> Sequence<MenuItem>> { fail() }
+
+    @Suppress("UNCHECKED_CAST")
+    fun <T: Any> add(type: Class<T>, items: Builder<T>.() -> Unit) {
+        mSingle.accumulate(type) { menu, item ->
+            Builder(menu, collectionUnwrap(item) as T).apply(items).asSequence()
+        }
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    fun <T: Any> addMany(type: Class<T>, items: Builder<Collection<T>>.() -> Unit) {
+        mMany.accumulate(type) { menu, item ->
+            Builder(menu, collectionWrap(item) as Collection<T>).apply(items).asSequence()
+        }
+    }
+
+    inline fun <reified T: Any> add(noinline items: Builder<T>.() -> Unit) = add(T::class.java, items)
+
+    inline fun <reified T: Any> addMany(noinline items: Builder<Collection<T>>.() -> Unit) = addMany(T::class.java, items)
+
+    operator fun get(contextMenu: ContextMenu, value: Any?): Sequence<MenuItem> {
+        val valueSingle = value?.let { collectionUnwrap(it) }
+        val valueMulti = value?.let { collectionWrap(value) }?.takeUnless { it.isEmpty() }
+
+        val items1Type = valueSingle?.javaClass ?: Void::class.java
+        val items1 = mSingle.getElementsOfSuperV(items1Type)
+
+        val itemsNType = valueMulti.asIf<Collection<*>>()?.getElementType()
+        val itemsN = itemsNType?.net { mMany.getElementsOfSuperV(it) } ?: listOf()
+
+        return (items1.asSequence()+itemsN.asSequence())
+                .map { it(contextMenu, value) }
+                .flatMap { sequenceOf(menuSeparator())+it }
+                .drop(1)
+    }
+
+    /** Allows DSL for [ContextMenuGenerator]. */
+    class Builder<T>(val contextMenu: ContextMenu, val selected: T) {
+
+        private val items = ArrayList<MenuItem>()
+
+        fun asSequence() = items.asSequence()
```

Comment:
shouldn't this methode be called `build` then?

Reply:
Any reason?
One implies what is returned, the other what it does. I suppose both are fine.

Reply:
was just a consideration, it can stay as it is

Reply:
Ok then it stays. For reference this is also employed by Kotlin, for example for directory traversal. There is an immutable builder that has asSequence() method instead of build().